src/dlfcn.c

```diff
@@ -467,7 +472,170 @@ char *dlerror( void )
+        optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].VirtualAddress == 0 )
+        return NULL;
+    *size = optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].Size;
+    return (IMAGE_IMPORT_DESCRIPTOR*)((BYTE*)module +
+        optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].VirtualAddress);
+}
+
+/*
+ * return symbol name for a given address
+ */
+static const char *get_symbol_name( HMODULE module, IMAGE_IMPORT_DESCRIPTOR *iid, void *addr )
+{
+    int i;
+    for(i = 0; iid[i].Characteristics != 0 && iid[i].FirstThunk != 0; i++) {
+        BYTE *base = (BYTE *)module; /* required to have correct calculations */
+        IMAGE_THUNK_DATA *thunkILT = (IMAGE_THUNK_DATA*)(iid[i].Characteristics + base);
+        IMAGE_THUNK_DATA *thunkIAT = (IMAGE_THUNK_DATA*)(iid[i].FirstThunk + base);
+        for(; thunkILT->u1.AddressOfData != 0; thunkILT++, thunkIAT++) {
+            if (IMAGE_SNAP_BY_ORDINAL(thunkILT->u1.Ordinal))
+              continue;
+            if ((void *)thunkIAT->u1.Function != addr)
+              continue;
+            IMAGE_IMPORT_BY_NAME *nameData = (IMAGE_IMPORT_BY_NAME*)(thunkILT->u1.AddressOfData + base);
+            return (const char *)nameData->Name;
+        }
+   }
+   return NULL;
+}
+
+static BOOL is_valid_address( void *addr )
+{
+    if( addr == NULL )
+        return FALSE;
+
+    /* check valid pointer */
+    MEMORY_BASIC_INFORMATION info;
+    SIZE_T result = VirtualQuery( addr, &info, sizeof( info ) );
+    if( result == 0 || info.AllocationBase == NULL || info.AllocationProtect == 0 || info.AllocationProtect == PAGE_NOACCESS )
+        return FALSE;
+    return TRUE;
+}
+
+/*
+ * Return state if address points to an import thunk
+ *
+ * An import thunk is setup with a 'jmp' instruction followed by an
+ * absolute address (32bit) or relative offset (64bit) pointing into
+ * the import address table (iat), which is partially maintained by
+ * the runtime linker.
+ */
+static BOOL is_import_thunk( void *addr )
+{
+    return *(short *)addr == 0x25ff ? TRUE : FALSE;
+}
+
+/*
+ * Return adress from the import address table (iat), if
+ * the original address points to a thunk table entry.
+ */
+static void *get_address_from_import_address_table( void *iat, DWORD iat_size, void *addr )
+{
+    BYTE *thkp = (BYTE*)addr;
+    /* get offset from thunk table (after instruction 0xff 0x25)
+     *   4018c8 <_VirtualQuery>: ff 25 4a 8a 00 00
+     */
+    ULONG offset = *(ULONG*)(thkp + 2);
+#ifdef _WIN64
+    /* On 64 bit the offset is relative
+     *   4018c8:	ff 25 4a 8a 00 00    	jmpq   *0x8a4a(%rip)        # 40a318 <__imp_VirtualQuery>
+     */
+    BYTE *ptr = (BYTE *)(thkp + 6 + offset);
+#else
+    /* On 32 bit the offset is absolute
+     *    4019b4:	ff 25 90 71 40 00    	jmp    *0x40719
+     */
+    BYTE *ptr = (BYTE *)offset;
+#endif
+
+    if( !is_valid_address( ptr ) ||  ptr < (BYTE *)iat || ptr > (BYTE *)iat + iat_size )
+        return NULL;

```

Comment:
> It looks like that current code for import thunk table in src/dlfcn.c does not work under MSVC :-(
> Tests which check it (those which are marked with PassWithDifferentAddress) are failing.

Now I found where is the problem and why tests are failing. Tests are correct and problem is really in this implementation. After I fixed problem with declaration of variables (as written before) I found that MSVC put function pointer outside of the IAT: `iat=0100307C iat_size=60 addr=01002508 ptr=0100102C`

So above check that `ptr` is inside of IAT is wrong, it really can be outside.

I changed code just to:
```c
    if( !is_valid_address( ptr ) )
        return NULL;
```
and tests started passing.

Reply:
To which function are you refering ?

Reply:
And does this happen with a 32bit or 64 bit application ? There are different calculations 
```
    ULONG offset = *(ULONG*)(thkp + 2);
#ifdef _WIN64
    /* On 64 bit the offset is relative
     *   4018c8:	ff 25 4a 8a 00 00    	jmpq   *0x8a4a(%rip)        # 40a318 <__imp_VirtualQuery>
     */
    BYTE *ptr = (BYTE *)(thkp + 6 + offset);
#else
    /* On 32 bit the offset is absolute
     *    4019b4:	ff 25 90 71 40 00    	jmp    *0x40719
     */
    BYTE *ptr = (BYTE *)offset;
#endif
```

Reply:
It happens on both 32bit and 64bit. See AppVeyor. Both 32bit and 64bit MSVC builds are failing due to this error.

I'm refering to get_address_from_import_address_table function, to code `if( !is_valid_address( ptr ) ||  ptr < (BYTE *)iat || ptr > (BYTE *)iat + iat_size )`. As wrote above it should be only `if( !is_valid_address( ptr ) )`.

Reply:
But there is no priblem in `#ifdef _WIN64` .. `#else` ... `#endif` and `BYTE *ptr` code. Problem is really only in that check after.

Reply:
According to my comments at https://github.com/dlfcn-win32/dlfcn-win32/pull/72#discussion_r546113857 for mingw it may be that msvc has a different binary layout. Would it be possible to get a assembler dump with symbols from the files libdl.lib and test-dladdr.exe for inspecting ?

Reply:
You can inspect it. Here are sources and i386 build via wdk: [dlfcn-win32-wdk-i386.tar.gz](https://github.com/dlfcn-win32/dlfcn-win32/files/5718382/dlfcn-win32-wdk-i386.tar.gz)

Reply:
The reason for this is that with this build libdl is added as a static library, which means that there are no import thunks and no iat entries for the functions that are normally exported by libdl. This also means that there can be no symbols for functions of a statically added library.

Reply:
> The reason for this is that with this build libdl is added as a static library, which means that there are no import thunks

But libdl is irrelevant here. There must be import thunk for GetModuleHandleA and VirtualQuery which are defined in test-dladdr.c file.

Reply:
And there really is an import thunk for `VirtualQuery`. See output from objdump.

Reply:
But this check 
```
result |= check_dladdr( (char*)VirtualQuery, "VirtualQuery", PassWithDifferentAddress );
```
is different.
```
iat=0100307C iat_size=60 addr=01002514 ptr=01001030
```
addr=01002514 points to the import thunk
```
1002514:	ff 25 30 10 00 01    	jmp    *0x1001030
```
This points to 0x1001030, which is inside the text segment
```
Disassembly of section .text:
01001000 <.text>:
 1001000:	7c 31                	jl     0x1001033
...
 100102f:	00 **52 32**             	add    %dl,0x32(%edx)
 1001032:	**00 00**                	add    %al,(%eax)
 1001034:	62 32                	bound  %esi,(%edx)
```
-> msvc places the import allocation table at the beginning of the text segment and not in a seperate code segment as with mingw. 

Therefore the check should be extended to allow the range of the text segment
```
        0x01001000 - 0x0100349a is .text
```

Reply:
More exactly from start of text segment to entry point
```
Entry point: 0x10027b3
0x01001000 - 0x0100349a is .text
``
which is 0x01001000 - 0x10027b3.

Reply:
I see...

This starting to be even more complicated.

And now I'm thinking, do we need to check if address is in IAT or text section? Is not enough if we check that address is valid via `VirtualQuery` function and just to know that address can be dereferenced and read?

Because now it is complicated and maybe other C compilers / linkers can put these addresses into other locations too.

Basically currently I do not see nothing which could cause process crash or any other issue if we omit this check (address is in IAT or text section). Or do you see something which can cause issue here?

Reply:
You can see the differences with objdump -x bin/test-dladdr.exe

mingw:
```
The Data Directory
...
Entry 1 00007000 00000618 Import Directory [parts of .idata]
...
Entry c 0000712c 000000dc Import Address Table Directory

There is an import table in .idata at 0x407000

The Import Tables (interpreted .idata section contents)
 vma:            Hint    Time      Forward  DLL       First
                 Table   Stamp     Chain    Name      Thunk
 00007000       00007050 00000000 00000000 00007530 0000712c
```

msvc:
```
The Data Directory
...
Entry 1 0000307c 0000003c Import Directory [parts of .idata]
...
Entry c 00001000 000000c4 Import Address Table Directory

There is an import table in .text at 0x100307c

The Import Tables (interpreted .text section contents)
 vma:            Hint    Time      Forward  DLL       First
                 Table   Stamp     Chain    Name      Thunk
 0000307c       000030b8 00000000 00000000 00003278 00001000
```

mingw:
Entry 1 **00007000** 00000618 Import Directory [parts of .idata]
Entry c **0000712c** 000000dc Import Address Table Directory

msvc:
Entry 1 **0000307c** 0000003c Import Directory [parts of .idata]
Entry c **00001000** 000000c4 Import Address Table Directory

The difference is that with mingw the Import Directory and the Import Address Table Directory are in one linear area, while with msvc there are two independent blocks. To fix that ptr need to be verified against the **Import Address Table Directory**, not Import Directory.

Reply:
Good catch! So it was wrong since beginning and due to linear area in mingw, this issue was perfectly hidden.